ansible_rulebook/engine.py

```diff
@@ -293,32 +298,29 @@ async def run_actions(self, data: Dict):
             finally:
                 logger.debug(lang.get_pending_events(self.name))
 
-            while not self.ruleset_queue_plan.plan.queue.empty():
-                matched = True
-                item = cast(
-                    ActionContext,
-                    await self.ruleset_queue_plan.plan.queue.get(),
-                )
-                result = await self.call_action(*item)
-                results.append(result)
-
-            if matched:
-                # TODO: how useful to log the results without logging the
-                # original event data?
-                await self.event_log.put(
-                    dict(type="ProcessedEvent", results=results)
-                )
-        except MessageNotHandledException:
-            logger.info("MessageNotHandledException: %s", data)
-            # TODO: is it really necessary to add such event to event_log?
-            await self.event_log.put(dict(type="MessageNotHandled"))
-        except ShutdownException:
-            await self.ruleset_queue_plan.source_queue.put(Shutdown())
-        except Exception:
-            logger.exception("Error processing %s", data)
-
-    async def call_action(
+    async def _drain_actionplan_queue(self):
+        logger.info("Waiting for actions on events from %s", self.name)
+        action_id = 0
+        while not self.stopped:
+            queue_item = await self.ruleset_queue_plan.plan.queue.get()
+            if isinstance(queue_item, Shutdown):
+                break
+            # TODO: consider uncomment the following but it will fail a lot of
+            # spec tests because Shutdown is issued immediately after the last
+            # event. Spec tests assume every thing is processed without lossing
+            # if self.stopped:
+            #    break
+
+            action_item = cast(ActionContext, queue_item)
+            action_id += 1
+            self.action_tasks[action_id] = asyncio.create_task(
+                self._call_action(action_id, *action_item)
+            )
+        logger.info("Stopped waiting for actions on events from %s", self.name)
```

Comment:
@bzwei Since this task runs independently it might make sense to put it in a try/except block and catch the asyncio.CancelledError to log a clean shutdown when a task is cancelled.